File: cpp/horner.cpp
/*
 * horner.cpp
 */


#include <iostream>

using namespace std;

// W(x) = 2x^3 + 3x^2 + 5x + 4 (8)
// W(x) = x (2x^2 + 3x + 5) + 4
// W(x) = x ( x (2x+3) + 5) + 4 (3)


float horner_it(int k, float wspl[], float x){
    float wynik = wspl[0];
    for (int i = 1; i < k+1; i++){
    wynik = wynik * x + wspl[i];
    }
    return wynik;
}


int main(int argc, char **argv)
{   float wspl[4];
    float x;
    int stopien = 3;
    
    cout<<"Podaj argument "<<endl;
    cin>> x;
    
    int i;
    
    for(i=0; i<4 ;i++) {
       cout << "Podaj współczynniki: ";
       cin >> wspl[4];
    }

	cout << "Wartość wielomianu: " << horner_it(stopien, wspl, x) << endl;
	return 0;
}

File: python/horner.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  horner.py
#


def horner_it(k, tbwsp, x):
    wynik = tbwsp[0]
    for i in range(1, k + 1):
        wynik = wynik * x + tbwsp[i]

    return wynik


def horner_rek(k, tbwsp, x):
    if k == 0:
        return tbwsp[0] * x
    return horner_rek(k - 1, tbwsp, x) * x + tbwsp[k]


def main(args):
    tbwsp = []
    stopien = 3
    x = int(input("Podaj argument: "))
    for i in range(0, 4):
        tmp = int(input("Podaj współczynnik: "))
        tbwsp.append(tmp)

    print("Wynik ", horner_rek(stopien, tbwsp, x))

    return 0


if __name__ == '__main__':
    import sys
sys.exit(main(sys.argv))
